packages/amplify-mqtt-server/src/server.ts

```diff
@@ -0,0 +1,290 @@
+
+    this.once('ready', () => {
+      this.callback(null, this);
+    });
+
+    this.on('ready', () => {
+      this.listener.subscribe('$SYS/+/new/clients', (topic, payload) => {
+        const serverId = topic.split('/')[1];
+        const clientId = payload;
+
+        if (this.clients[clientId] && serverId !== this.id) {
+          this.clients[clientId].close(null, 'new connection request');
+        }
+      });
+    });
+
+    if (this.options.publishSubscriptions) {
+      this.on('subscribed', (topic, client) => {
+        this.publish({
+          topic: '$SYS/' + this.id + '/new/subscribes',
+          payload: JSON.stringify({
+            clientId: client.id,
+            topic: topic,
+          }),
+        });
+      });
+
+      this.on('unsubscribed', (topic, client) => {
+        this.publish({
+          topic: '$SYS/' + this.id + '/new/unsubscribes',
+          payload: JSON.stringify({
+            clientId: client.id,
+            topic: topic,
+          }),
+        });
+      });
+    }
+
+    this.on('clientDisconnected', client => {
+      if (this.options.publishClientDisconnect) {
+        this.publish({
+          topic: '$SYS/' + this.id + '/disconnect/clients',
+          payload: client.id,
+        });
+      }
+      delete this.clients[client.id];
+    });
+  }
+
+  toString() {
+    return 'AmplifyMQTTServer.Server';
+  }
+
+  subscribe(topic, callback, done) {
+    this.listener.subscribe(topic, callback, done);
+  }
+
+  publish(packet, client?, callback?) {
+    let logger = this.logger;
+
+    if (typeof client === 'function') {
+      callback = client;
+      client = null;
+    } else if (client) {
+      logger = client.logger;
+    }
+
+    if (!callback) {
+      callback = nop;
+    }
+
+    const newPacket = {
+      topic: packet.topic,
+      payload: packet.payload,
+      messageId: this.generateUniqueId(),
+      qos: packet.qos,
+      retain: packet.retain,
+    };
+
+    const opts: any = {
+      qos: packet.qos,
+      messageId: newPacket.messageId,
+    };
+
+    if (client) {
+      opts.clientId = client.id;
+    }
+
+    if (this.closed) {
+      logger.debug({ packet: newPacket }, 'not delivering because we are closed');
+      return;
+    }
+
+    this.listener.publish(newPacket.topic, newPacket.payload, opts, () => {
+      if (newPacket.topic.indexOf('$SYS') >= 0) {
+        logger.trace({ packet: newPacket }, 'published packet');
+      } else {
+        logger.debug({ packet: newPacket }, 'published packet');
+      }
+      this.emit('published', newPacket, client);
+      callback(undefined, newPacket);
+    });
+  }
+
+  close(callback = nop) {
+    const stuffToClose = [];
+
+    if (this.closed) {
+      return callback();
+    }
+
+    this.closed = true;
+
+    Object.keys(this.clients).forEach(i => {
+      stuffToClose.push(this.clients[i]);
+    });
+
+    steed.each(
+      stuffToClose,
+      (toClose, cb) => {
+        toClose.close(cb, 'server closed');
```

Comment:
What do you think about putting the ```close``` invocation into a try catch to make sure that if 1 close fails, then we can still call the others and not bailing out from the loop.